src/MatrixUtilities.cpp

```diff
@@ -964,85 +964,173 @@ void GenerateOffsets(MPI_Comm comm, int local_size, mfem::Array<HYPRE_Int>& offs
-    A_.Swap(*tmp);
+    mfem::SparseMatrix DMinvDT = smoothg::Mult(D, MinvDT_);
+    A_.Swap(DMinvDT);
 
     // Eliminate the first unknown so that A_ is invertible
     A_.EliminateRowCol(0);
     solver_ = make_unique<mfem::UMFPackSolver>(A_);
 }
 
-void LocalGraphEdgeSolver::Mult(const mfem::Vector& rhs,
-                                mfem::Vector& sol_sigma)
+void LocalGraphEdgeSolver::Init(const mfem::SparseMatrix& M, const mfem::SparseMatrix& D)
 {
-    // Set rhs(0)=0 so that the modified system after
-    // the elimination is consistent with the original one
-    mfem::Vector& rhs_copy = const_cast<mfem::Vector&>(rhs);
-    double rhs_0 = rhs_copy(0);
-    rhs_copy(0) = 0.;
+    offsets_.SetSize(3);
+    offsets_[0] = 0;
+    offsets_[1] = M.Height();
+    offsets_[2] = M.Height() + D.Height();
 
-    mfem::Vector sol_u_tmp(A_.Size());
-    sol_u_tmp = 0.;
-    solver_->Mult(rhs_copy, sol_u_tmp);
+    mfem::SparseMatrix D_copy(D, false);
+    D_copy.EliminateRow(0);
+    mfem::SparseMatrix DT = smoothg::Transpose(D_copy);
 
-    // Set rhs(0) back to its original vale (rhs is const Vector)
-    rhs_copy(0) = rhs_0;
+    mfem::SparseMatrix W(D.Height(), D.Height());
+    W.Add(0, 0, 1.0);
+    W.Finalize();
 
-    // SparseMatrix::Mult asserts that sol.Size() should equal MinvDT_->Height()
-    MinvDT_.Mult(sol_u_tmp, sol_sigma);
+    mfem::BlockMatrix block_A(offsets_);
+    block_A.SetBlock(0, 0, const_cast<mfem::SparseMatrix*>(&M));
+    block_A.SetBlock(1, 0, &D_copy);
+    block_A.SetBlock(0, 1, &DT);
+    block_A.SetBlock(1, 1, &W);
+
+    unique_ptr<mfem::SparseMatrix> tmp_A(block_A.CreateMonolithic());
+    A_.Swap(*tmp_A);
+
+    solver_ = make_unique<mfem::UMFPackSolver>(A_);
+
+    rhs_ = make_unique<mfem::BlockVector>(offsets_);
+    sol_ = make_unique<mfem::BlockVector>(offsets_);
+    rhs_->GetBlock(0) = 0.0;
 }
 
-void LocalGraphEdgeSolver::Mult(const mfem::Vector& rhs,
-                                mfem::Vector& sol_sigma, mfem::Vector& sol_u)
+void LocalGraphEdgeSolver::Mult(const mfem::Vector& rhs, mfem::Vector& sol_sigma) const
 {
     // Set rhs(0)=0 so that the modified system after
     // the elimination is consistent with the original one
     mfem::Vector& rhs_copy = const_cast<mfem::Vector&>(rhs);
     double rhs_0 = rhs_copy(0);
     rhs_copy(0) = 0.;
 
-    sol_u = 0.;
-    solver_->Mult(rhs_copy, sol_u);
-    orthogonalize_from_constant(sol_u);
+    if (M_is_diag_)
+    {
+        mfem::Vector sol_u_tmp(rhs.Size());
+        solver_->Mult(rhs, sol_u_tmp);
+        MinvDT_.Mult(sol_u_tmp, sol_sigma);
+    }
+    else
+    {
+        rhs_->GetBlock(1) = rhs_copy;
+        solver_->Mult(*rhs_, *sol_);
+        sol_sigma = sol_->GetBlock(0);
+    }
 
-    // Set rhs(0) back to its original vale (rhs is const Vector)
+    // Set rhs_block1(0) back to its original vale (rhs is const BlockVector)
     rhs_copy(0) = rhs_0;
+}
+
+void LocalGraphEdgeSolver::Mult(const mfem::Vector& rhs0, const mfem::Vector& rhs1,
```

Comment:
Could change `rhs0` -> `rhs_sigma` etc. here, but it's less important than in the .hpp file.